File: samples cs/ShotgunSurgeryExample.cs
using System;
using System.Collections.Generic;

/// <summary>
/// PricingRule and dependant domain classes
/// </summary>

// --- PricingRule ---

public class PricingRule
{
    public string RuleId { get; }
    public string RuleType { get; }      // "flat", "percentage", "tiered"
    public double BaseRate { get; }
    public Dictionary<string, double> TierRates { get; }
    public bool ApplyToSubtotal { get; }
    public string Currency { get; }

    public PricingRule(string ruleId, string ruleType, double baseRate)
    {
        RuleId = ruleId;
        RuleType = ruleType;
        BaseRate = baseRate;
        TierRates = new Dictionary<string, double>();
        ApplyToSubtotal = true;
        Currency = "USD";
    }

    public double ComputeAdjustment(double amount)
    {
        return RuleType switch
        {
            "flat"       => BaseRate,
            "percentage" => amount * BaseRate,
            "tiered"     => ComputeTieredAdjustment(amount),
            _            => 0.0
        };
    }

    private double ComputeTieredAdjustment(double amount)
    {
        double adjustment = 0;
        foreach (var tier in TierRates)
        {
            double threshold = double.Parse(tier.Key);
            if (amount >= threshold)
            {
                adjustment = tier.Value * amount;
            }
        }
        return adjustment;
    }
}


// --- OrderService ---

public class OrderService
{
    private List<PricingRule> applicableRules;

    public double CalculateTotal(List<LineItem> items)
    {
        double subtotal = 0;
        foreach (var item in items)
        {
            subtotal += item.Price * item.Quantity;
        }

        double discount = 0;
        foreach (var rule in applicableRules)
        {
            if (rule.ApplyToSubtotal)
            {
                discount += rule.ComputeAdjustment(subtotal);
            }
            else
            {
                foreach (var item in items)
                {
                    discount += rule.ComputeAdjustment(item.Price);
                }
            }
        }

        return subtotal - discount;
    }
}

// --- InvoiceService ---

public class InvoiceService
{
    public string GenerateInvoice(Order order, List<PricingRule> rules)
    {
        var invoice = new System.Text.StringBuilder();
        invoice.AppendLine($"Invoice for Order: {order.Id}");

        foreach (var rule in rules)
        {
            invoice.AppendLine(
                $"Applied rule: {rule.RuleId} ({rule.RuleType}) rate={rule.BaseRate} currency={rule.Currency}");
        }

        return invoice.ToString();
    }
}

// --- DiscountEngine ---

public class DiscountEngine
{
    private List<PricingRule> allRules;

    public List<PricingRule> SelectApplicableRules(string customerTier, double orderAmount)
    {
        var applicable = new List<PricingRule>();

        foreach (var rule in allRules)
        {
            // Decision logic tightly coupled to PricingRule's internal RuleType values
            if (rule.RuleType == "percentage" && orderAmount > 100)
                applicable.Add(rule);
            else if (rule.RuleType == "tiered" && rule.TierRates.Count > 0)
                applicable.Add(rule);
            else if (rule.RuleType == "flat")
                applicable.Add(rule);
        }

        return applicable;
    }
}

// --- TaxCalculator ---

public class TaxCalculator
{
    private readonly double _taxRate;

    public TaxCalculator(double taxRate) { _taxRate = taxRate; }

    public double ComputeTax(double subtotal, List<PricingRule> discountRules)
    {
        double afterDiscount = subtotal;

        foreach (var rule in discountRules)
        {
            double adjustment = rule.ComputeAdjustment(afterDiscount);
            afterDiscount -= adjustment;
        }

        return afterDiscount * _taxRate;
    }
}

// --- QuoteGenerator ---

public class QuoteGenerator
{
    public string BuildQuote(List<LineItem> items, List<PricingRule> rules)
    {
        var quote = new System.Text.StringBuilder();
        quote.AppendLine("Price Quote");
        quote.AppendLine("===========");

        double subtotal = 0;
        foreach (var item in items)
        {
            subtotal += item.Price * item.Quantity;
        }

        quote.AppendLine($"Subtotal: {subtotal}");

        foreach (var rule in rules)
        {
            double adj = rule.ComputeAdjustment(subtotal);
            quote.AppendLine($"Rule {rule.RuleId} ({rule.RuleType}): -{adj} {rule.Currency}");
        }

        return quote.ToString();
    }
}

// --- BillingService ---

public class BillingService
{
    public double ProcessBilling(Customer customer, List<PricingRule> rules)
    {
        double baseCharge = customer.SubscriptionAmount;

        foreach (var rule in rules)
        {
            if (rule.ApplyToSubtotal)
            {
                double discount = rule.ComputeAdjustment(baseCharge);
                baseCharge -= discount;
            }
        }

        // Log the applied pricing
        foreach (var rule in rules)
        {
            Console.WriteLine(
                $"Applied {rule.RuleType} rule {rule.RuleId} with rate {rule.BaseRate}");
        }

        return baseCharge;
    }
}


// --- Supporting classes ---

public class LineItem
{
    public string ProductId { get; set; }
    public double Price { get; set; }
    public int Quantity { get; set; }
}

public class Order
{
    public string Id { get; set; }
    public List<LineItem> Items { get; set; }
}

public class Customer
{
    public string Name { get; set; }
    public double SubscriptionAmount { get; set; }
}

File: samples/ShotgunSurgeryExample.java
package com.example.samples;

import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.HashMap;

/** PricingRule and dependant domain classes */

// --- PricingRule ---

class PricingRule {
    private String ruleId;
    private String ruleType;    // "flat", "percentage", "tiered"
    private double baseRate;
    private Map<String, Double> tierRates;
    private boolean applyToSubtotal;
    private String currency;

    public PricingRule(String ruleId, String ruleType, double baseRate) {
        this.ruleId = ruleId;
        this.ruleType = ruleType;
        this.baseRate = baseRate;
        this.tierRates = new HashMap<>();
        this.applyToSubtotal = true;
        this.currency = "USD";
    }

    public String getRuleId() { return ruleId; }
    public String getRuleType() { return ruleType; }
    public double getBaseRate() { return baseRate; }
    public Map<String, Double> getTierRates() { return tierRates; }
    public boolean isApplyToSubtotal() { return applyToSubtotal; }
    public String getCurrency() { return currency; }

    public double computeAdjustment(double amount) {
        switch (ruleType) {
            case "flat":
                return baseRate;
            case "percentage":
                return amount * baseRate;
            case "tiered":
                return computeTieredAdjustment(amount);
            default:
                return 0.0;
        }
    }

    private double computeTieredAdjustment(double amount) {
        double adjustment = 0;
        for (Map.Entry<String, Double> tier : tierRates.entrySet()) {
            double threshold = Double.parseDouble(tier.getKey());
            if (amount >= threshold) {
                adjustment = tier.getValue() * amount;
            }
        }
        return adjustment;
    }
}


// --- OrderService ---

class OrderService {
    private List<PricingRule> applicableRules;

    public double calculateTotal(List<LineItem> items) {
        double subtotal = 0;
        for (LineItem item : items) {
            subtotal += item.getPrice() * item.getQuantity();
        }

        double discount = 0;
        for (PricingRule rule : applicableRules) {
            if (rule.isApplyToSubtotal()) {
                discount += rule.computeAdjustment(subtotal);
            } else {
                // Apply per-item
                for (LineItem item : items) {
                    discount += rule.computeAdjustment(item.getPrice());
                }
            }
        }

        return subtotal - discount;
    }
}

// --- InvoiceService ---

class InvoiceService {
    public String generateInvoice(Order order, List<PricingRule> rules) {
        StringBuilder invoice = new StringBuilder();
        invoice.append("Invoice for Order: ").append(order.getId()).append("\n");

        for (PricingRule rule : rules) {
            invoice.append("Applied rule: ").append(rule.getRuleId());
            invoice.append(" (").append(rule.getRuleType()).append(")");
            invoice.append(" rate=").append(rule.getBaseRate());
            invoice.append(" currency=").append(rule.getCurrency()).append("\n");
        }

        return invoice.toString();
    }
}

// --- DiscountEngine ---

class DiscountEngine {
    private List<PricingRule> allRules;

    public List<PricingRule> selectApplicableRules(String customerTier, double orderAmount) {
        List<PricingRule> applicable = new ArrayList<>();

        for (PricingRule rule : allRules) {
            // Decision logic tightly coupled to PricingRule's internal ruleType values
            if (rule.getRuleType().equals("percentage") && orderAmount > 100) {
                applicable.add(rule);
            } else if (rule.getRuleType().equals("tiered") && !rule.getTierRates().isEmpty()) {
                applicable.add(rule);
            } else if (rule.getRuleType().equals("flat")) {
                applicable.add(rule);
            }
        }

        return applicable;
    }
}

// --- TaxCalculator ---

class TaxCalculator {
    private double taxRate;

    public TaxCalculator(double taxRate) {
        this.taxRate = taxRate;
    }

    public double computeTax(double subtotal, List<PricingRule> discountRules) {
        double afterDiscount = subtotal;

        for (PricingRule rule : discountRules) {
            double adjustment = rule.computeAdjustment(afterDiscount);
            if (rule.getRuleType().equals("flat")) {
                afterDiscount -= adjustment;
            } else {
                afterDiscount -= adjustment;
            }
        }

        return afterDiscount * taxRate;
    }
}

// --- QuoteGenerator ---

class QuoteGenerator {
    public String buildQuote(List<LineItem> items, List<PricingRule> rules) {
        StringBuilder quote = new StringBuilder();
        quote.append("Price Quote\n");
        quote.append("===========\n");

        double subtotal = 0;
        for (LineItem item : items) {
            subtotal += item.getPrice() * item.getQuantity();
        }

        quote.append("Subtotal: ").append(subtotal).append("\n");

        for (PricingRule rule : rules) {
            double adj = rule.computeAdjustment(subtotal);
            quote.append("Rule ").append(rule.getRuleId());
            quote.append(" (").append(rule.getRuleType()).append("): -");
            quote.append(adj).append(" ").append(rule.getCurrency()).append("\n");
        }

        return quote.toString();
    }
}

// --- BillingService ---

class BillingService {
    public double processBilling(Customer customer, List<PricingRule> rules) {
        double baseCharge = customer.getSubscriptionAmount();

        for (PricingRule rule : rules) {
            if (rule.isApplyToSubtotal()) {
                double discount = rule.computeAdjustment(baseCharge);
                baseCharge -= discount;
            }
        }

        // Log the applied pricing
        for (PricingRule rule : rules) {
            System.out.println("Applied " + rule.getRuleType() + " rule " +
                             rule.getRuleId() + " with rate " + rule.getBaseRate());
        }

        return baseCharge;
    }
}


// --- Supporting classes ---

class LineItem {
    private String productId;
    private double price;
    private int quantity;

    public String getProductId() { return productId; }
    public double getPrice() { return price; }
    public int getQuantity() { return quantity; }
}

class Order {
    private String id;
    private List<LineItem> items;

    public String getId() { return id; }
    public List<LineItem> getItems() { return items; }
}

class Customer {
    private String name;
    private double subscriptionAmount;

    public String getName() { return name; }
    public double getSubscriptionAmount() { return subscriptionAmount; }
}
